cmd/kubeadm/app/cmd/util/cmdutil.go

```diff
@@ -62,15 +62,16 @@ func ValidateExactArgNumber(args []string, supportedArgs []string) error {
 	return nil
 }
 
-// FindExistingKubeConfig returns the localtion of kubeconfig
-func FindExistingKubeConfig(file string) string {
-	// The user did provide a --kubeconfig flag. Respect that and threat it as an
-	// explicit path without building a DefaultClientConfigLoadingRules object.
-	if file != kubeadmconstants.GetAdminKubeConfigPath() {
-		return file
+// FindExistingKubeConfig returns the location of a kubeconfig file.
+// If the user passed the --kubeconfig flag respect it, otherwise search for existing
+// kubeconfig files in default locations.
+func FindExistingKubeConfig(flags *pflag.FlagSet) string {
```

Comment:
if we check over a string, there is no way to determine if the user set the flag or if it was defaulted to a value by cobra.

in the linked issue the user was passing: `--kubeconfig=/etc/kubernetes/admin.conf` and it still used the KUBECONFIG env.

flags.Changed() solves this problem.

Reply:
We must not default the flag in Cobra. We have to make sure it's defaulted to an empty string.

Reply:
This is kind of the case of the `cri-socket` flag.

Reply:
here is an example of what happens in kubeoconfig.go:
https://github.com/kubernetes/kubernetes/blob/639dae9a9a6602ac194a171744cf84bdcba9f29a/cmd/kubeadm/app/cmd/config.go#L65-L84
- we default a local variable to `/etc/kubernetes/....`
- then use `options..` to add the flag.
- at this point the default becomes `/etc/kubernetes/...`

how should this be handled instead?
the CRI flag passes "" as the default.
https://github.com/kubernetes/kubernetes/blob/b5b627d522ed74549e2583b258a7df0e3ffd33ea/cmd/kubeadm/app/cmd/reset.go#L88

problem is we want to show `/etc/kubernetes...` as the default path when the user types `kubeadm... --help`.

Reply:
`pflag` supports special field in its `Flag` struct called `DefValue` ([see here](https://github.com/spf13/pflag/blob/24fa6976df40757dce6aea913e7b81ade90530e1/flag.go#L176)). It's used solely for printing the default value for usage information and not for initialization. You can overwrite that just after flag insertion with something along those lines:

```
func AddKubeConfigFlag(fs *pflag.FlagSet, kubeConfigFile *string, defaultKubeConfigFile string) {
	fs.StringVar(kubeConfigFile, KubeconfigPath, "", "The kubeconfig file to use when talking to the cluster. If the flag is not set, a set of standard locations are searched for an existing KubeConfig file.")
	fs.Lookup(KubeconfigPath).DefValue = defaultKubeConfigFile
}
```

Then the invocation will be something along those lines:
```
kubeConfigPath := ""
// ....
options.AddKubeConfigFlag(cmd.Flags(), &kubeConfigPath, constants.GetAdminKubeConfigPath())
```

In theory this should work, but I did not test it.

P.S. You need the `Lookup` call, because `StringVar` does not return a `*pflag.Flag`.

Reply:
> `pflag` supports special field in its `Flag` struct called `DefValue`

ok i didn't know this existed.
will give this a test in the following days.